[WIP] Implement support for volume group mirroring

Madhu-1 opened the pull request:
This PR does the required changes so that the volume group can also implement the Mirror interface and we do need to make many changes in the replication code to work differently for the mirroring of an image or a mirroring of a group.

Note:- The last commit to add use go-ceph is commented out, please don't review it, Will remove that code and make changes to the mirror implementation of the group. except that other commits can be reviewed.

Depends-on: ceph/go-ceph#1010

nixpanic commented on internal/rbd/manager.go:
```diff
@@ -124,7 +124,19 @@ func (mgr *rbdManager) GetVolumeByID(ctx context.Context, id string) (types.Volu
 	volume, err := GenVolFromVolID(ctx, id, creds, mgr.secrets)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get volume from id %q: %w", id, err)
+		switch {
```
this commit (and maybe others) can be put in a separate PR for earlier merging and keeping follow-ups smaller if you like

Madhu-1 replied:
Let me see if i can do that 👍🏻

nixpanic commented on internal/rbd/manager.go:
```diff
@@ -295,3 +295,43 @@ func (mgr *rbdManager) DeleteVolumeGroup(ctx context.Context, vg types.VolumeGro
+	if rep == nil {
+		rbdVol, err := mgr.GetVolumeByID(ctx, volumeID)
+		if err != nil {
+			return nil, nil, err
```
consider adding a description to the returned `err`

nixpanic commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -117,6 +117,21 @@ func (vs *VolumeGroupServer) CreateVolumeGroup(
+	// extract the flatten mode
+	flattenMode, err := getFlattenMode(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
```
This needs to return a grpc error. `getFlattenMode` is not appropriate to do so.

Madhu-1 replied:
getFlattenMode is returning grpc error same is already used in replication.

nixpanic commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -343,17 +343,40 @@ func (vs *VolumeGroupServer) ModifyVolumeGroupMembership(
+	}
+
+	// extract the flatten mode
+	flattenMode, err := getFlattenMode(ctx, req.GetParameters())
```
same here, `getFlattenMode()` probably does not return a grpc error.

nixpanic commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -343,17 +343,40 @@ func (vs *VolumeGroupServer) ModifyVolumeGroupMembership(
+	for _, vol := range volumes {
+		err = vol.HandleParentImageExistence(ctx, flattenMode)
+		if err != nil {
+			return nil, getGRPCError(err)
```
consider adding more details to the error, include at least the volume name/id

Madhu-1 replied:
let me check that if HandleParentImageExistence already including it

nixpanic commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -304,7 +304,7 @@ func (rs *ReplicationServer) EnableVolumeReplication(ctx context.Context,
 	}
 
-	info, err := mirror.GetMirroringInfo()
+	info, err := mirror.GetMirroringInfo(ctx)
```
can be in an other PR too

nixpanic commented on internal/rbd/group/volume_group.go:
```diff
@@ -151,18 +151,17 @@ func GetVolumeGroup(
 
-// String makes it easy to include the volumeGroup object in log and error
-// messages.
+// String returns the image-spec (pool/{namespace}/{name}) format of the group.
```
can be in an other PR too

nixpanic commented on internal/rbd/group/volume_group.go:
```diff
@@ -465,3 +470,300 @@ func (vg *volumeGroup) RemoveVolume(ctx context.Context, vol types.Volume) error
+
+// GroupStatus is a wrapper around librbd.MirrorGroupInfo that contains the
+// group mirror info.
+type GroupStatus struct {
```
why not just do it like
```go
type GroupStatus *librbd.MirrorGroupInfo
```

Madhu-1 replied:
>Methods defined on the original type are not available to the new type unless you explicitly defined

if we create a type i need to implement duplication methods of MirrorGroupInfo here which is again not much useful, lets keep it this way which allow us to use methods of librbd struct without any problem

Madhu-1 commented:
@nixpanic This is ready for one more round of review :)

nixpanic commented on internal/rbd/types/manager.go:
```diff
@@ -47,4 +49,8 @@ type Manager interface {
+
+	// GetMirrorSource returns the source of the mirror for the given volume or group.
+	GetMirrorSource(ctx context.Context, volumeID string,
+		rep *replication.ReplicationSource) ([]Volume, Mirror, error)
```
it is not clear to me why `[]Volume` is returned. Should that not be either a single Volume or a VolumeGroup?

nixpanic replied:
would it make sense to have a `Mirror.GetSource()` function so that the API is easier to understand?

Rakshith-R replied:
Let's put everything related to processing volumes within functions in mirror type ?

Madhu-1 replied:
>would it make sense to have a Mirror.GetSource() function so that the API is easier to understand?

@nixpanic I put it with the Manager because that is the one who is generating the volume and group from an ID, moving this to Mirroring

>it is not clear to me why []Volume is returned. Should that not be either a single Volume or a VolumeGroup?

This is required because we need to work on all the volumes if its a group, i am trying to avoid type casting and doing the ListVolumes and generating the volumes again that contains the connections, let me know if you have any suggestions

> Let's put everything related to processing volumes within functions in mirror type ?

I mostly added all the mirroring specific things to mirror type but there are again generic one as well that the reason i left it in volume type as it makes more sense.

Rakshith-R replied:
:+1: 
We can check on volumes and refactor it later on.
Everything LGTM
Only concern now is ceph's API stability and go-ceph pr going in.

https://github.com/ceph/go-ceph/pull/1010#issuecomment-2223542910
https://github.com/ceph/ceph/pull/53793#discussion_r1670404359

nixpanic commented on internal/rbd/group/volume_group.go:
```diff
@@ -469,3 +474,296 @@ func (vg *volumeGroup) ListVolumes(ctx context.Context) ([]types.Volume, error)
 	return vg, nil
 }
+
+func (vg *volumeGroup) EnableMirroring(ctx context.Context, mode librbd.ImageMirrorMode) error {
```
I wonder if it is worth to simplify things a little by placing this in a separate file `group_mirror.go`?

```go
type volumeGroupMirror *volumeGroup // or make it a struct with extra members?

func (vgm *volumeGroupMirror) EnableMirroring(ctx context.Context, mode librbd.ImageMirrorMode) error {
    ...
```

This would keep the VolumeGroup interface smal(ler). I am considering a similar approach for VolumeGroupSnapshot.

nixpanic commented on internal/rbd/group/volume_group.go:
```diff
@@ -469,3 +474,296 @@ func (vg *volumeGroup) ListVolumes(ctx context.Context) ([]types.Volume, error)
+		return nil, fmt.Errorf("failed to get volume group mirroring info %q: %w", vg, err)
+	}
+
+	return GroupInfo{MirrorGroupInfo: info}, nil
```
you could also do

```go
type GroupInfo *librbd.MirrorGroupInfo
```

and cast it here

```go
return GroupInfo(info)
```

That looks a little simpler to me.

Madhu-1 replied:
Yes but again this will have problem like https://github.com/ceph/ceph-csi/pull/4739#discussion_r1697232084 and adding new field to GroupInfo need more refractoring, IMO this should be okay but the struct name should not be exported(taken care of it)

nixpanic commented on internal/rbd/group/volume_group.go:
```diff
@@ -469,3 +474,296 @@ func (vg *volumeGroup) ListVolumes(ctx context.Context) ([]types.Volume, error)
+}
+
+func (status GlobalMirrorGroupStatus) GetState() string {
+	return status.GlobalMirrorGroupStatus.Info.State.String()
```
could any of the members be `nil`?

Madhu-1 replied:
Not possible as the members are not pointers

nixpanic commented on internal/rbd/group/volume_group.go:
```diff
@@ -469,3 +474,296 @@ func (vg *volumeGroup) ListVolumes(ctx context.Context) ([]types.Volume, error)
+
+// GroupInfo is a wrapper around librbd.MirrorGroupInfo that contains the
+// group mirror info.
+type GroupInfo struct {
```
If you want to use a struct, consider exposing an interface instead of the struct. That encourages a cleaner API towards consumers.

Rakshith-R commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -280,15 +280,12 @@ func (rs *ReplicationServer) EnableVolumeReplication(ctx context.Context,
 	defer mgr.Destroy(ctx)
 
-	rbdVol, err := mgr.GetVolumeByID(ctx, volumeID)
+	volumes, mirror, err := mgr.GetMirrorSource(ctx, volumeID, req.GetReplicationSource())
```
let's call `volumeID` something else like reqID or replicationID.
It can then be either referring to a single volume or group (or later on snapshot).

Rakshith-R commented on internal/rbd/manager.go:
```diff
@@ -294,3 +295,98 @@ func (mgr *rbdManager) DeleteVolumeGroup(ctx context.Context, vg types.VolumeGro
+func (mgr *rbdManager) GetMirrorSource(ctx context.Context, volumeID string,
+	rep *replication.ReplicationSource,
+) ([]types.Volume, types.Mirror, error) {
+	// Backward compatibility: if rep is nil, we assume that the sidecar is still old and
+	// setting only volumeId not the replication source.
+	if rep == nil {
+		rbdVol, err := mgr.GetVolumeByID(ctx, volumeID)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get volume by id %q: %w", volumeID, err)
+		}
+		defer func() {
+			if err != nil {
+				rbdVol.Destroy(ctx)
+			}
+		}()
+		var mir types.Mirror
+		mir, err = rbdVol.ToMirror()
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to convert volume %s to mirror: %w", rbdVol, err)
+		}
+
+		return []types.Volume{rbdVol}, mir, nil
+	}
+	switch {
+	case rep.GetVolume() != nil:
```
```suggestion
	switch {
	// Backward compatibility: if rep is nil, we assume that the sidecar is still old and
	// set only volumeId, else rely on replicationSource.
	case rep == nil || rep.GetVolume() != nil:
```

Madhu-1 committed:
rbd: implement volume group using go-ceph

This adds the required functionality to
call the go-ceph API's for the rbd
volume group.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
ci: remove this commit later

This is temporary test only
commit and not suitable for
review.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: implement GetMirrorSource in manager

implementing GetMirrorSource in manager
to return volume or the volumegroup based
on the replication source, if replication
source is nil return the volume details
for backward compatibility.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
ci: extra commit will be dropped

This is extra commit will be dropped

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
ci: extra commit to fix panic

This commit will be removed later

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: delete group only if its primary

This is a work of Nikhil which need to be
applied on top of this PR to test the
feature.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>
